### PR TITLE
fix(aggregate): align exact AVG types with MySQL

### PR DESCRIPTION
**Pull request by daviszhen**
## What type of PR is this?

- [x] Bug

## What this PR does / why we need it

Continues #24436 by aligning the public `AVG` contract for exact numeric inputs with MySQL instead of applying a CTAS-only schema patch.

- Exact integer inputs now return DECIMAL with four fractional digits, e.g. `AVG(INT) -> DECIMAL(14,4)` and `AVG(BIGINT) -> DECIMAL(23,4)`.
- DECIMAL inputs use `precision + 4` and `scale + 4`, e.g. `AVG(DECIMAL(20,6)) -> DECIMAL(24,10)`.
- Results that exceed DECIMAL128 precision are promoted to DECIMAL256, capped at MySQL's precision/scale limits, so high-precision ordinary and window AVG no longer overflow in the Decimal128 accumulator.
- FLOAT/DOUBLE AVG remains DOUBLE.
- Ordinary aggregates, window aggregates, CTAS, and protocol/JDBC metadata share the same return-type rule.
- VARIANCE/STDDEV keep their existing return type and behavior.

This fixes both metadata and value semantics. For example, AVG over integer values `1, 2, 4` now returns `2.3333` rather than a DOUBLE approximation.

SUM precision rules are intentionally out of scope and will be handled separately.

## Which issue(s) this PR fixes

Related to #24436

## Special notes for your reviewer

The Decimal128-to-Decimal256 promotion is required for inputs such as `DECIMAL(38,19)`: clamping the result to DECIMAL128 would either discard integer capacity or lose the required four additional fractional digits.

## Testing

- `make build`
- `go test ./pkg/sql/colexec/aggexec -count=1`
- `go test ./pkg/sql/plan ./pkg/sql/colexec/window -count=1`
- 23 affected distributed BVT case files with current mo-tester (`-g -n`), including `func_aggr_avg`, numeric/decimal/decimal256, prepared aggregates, grouping, DISTINCT, CTAS, and the full window case
- Local mo-service checks for ordinary AVG, window AVG, CTAS schema, and `DECIMAL(38,19)` promotion

### CODE REVIEW

**Comment by daviszhen** [comment]
已更新分支并处理 failing checks 与代码冲突：

- 修复 YEAR AVG 使用 `types.MoYear` 向量却按 `int16` 读取导致的类型检查 panic；同时保留 YEAR SUM 的 int64 累加路径。
- fulltext BM25 的内部 `AVG(pos)` 显式 CAST 为 DOUBLE，保持 `runCountStar` 的 float64 读取契约。
- 更新 `func_arrayagg` 的 AVG DECIMAL 结果精度。
- 合并最新 `upstream/main`，并按当前 Decimal128 溢出用例解决 `window.result` 冲突。

本机定向 Go 测试受 macOS 原生依赖头文件缺失影响，已由 CI 重新验证。

**Comment by daviszhen** [comment]
Addressed in commit adde9798bc. Reconciled the #27869 overflow fixtures with the exact AVG contract: DECIMAL(38,10) now uses the required DECIMAL(42,14) Decimal128-to-Decimal256 path, so promoted boundary cases and window/BVT expectations are successful results rather than obsolete Decimal128 overflow errors. Updated the PK-group-elimination assertions accordingly, retained narrow-target decAvg error coverage, and kept Decimal256 declared-precision overflow coverage.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
Deep-reviewed exact head `adde9798bc690853fa102394907551758a477824`. The current CI is green, but I found four blockers that the existing tests do not cover.

1. **[P1 correctness] AVG is rounded twice, which can change the public value.**

   In `decAvg` (`sumavg2.go:1291-1297` and `1312-1318`), `Decimal*.Div` first rounds to its internal division scale and the result is then rounded again by `Scale(resultType.Scale-scale)`. Exact-head probes reproduce both paths:

   - integer AVG: one `1` plus 112 zeroes returns `0.0089`; directly rounding exact `1/113 = 0.008849...` to scale 4 is `0.0088`;
   - decimal AVG: one `0.01` plus 112 zeroes returns `0.000089`; directly rounding exact `0.01/113 = 0.000088495...` to scale 6 is `0.000088`.

   MySQL's implementation divides directly at `div_precision_increment`, rather than rounding at one scale and rescaling again: https://github.com/mysql/mysql-server/blob/trunk/sql/item_sum.cc. Please make finalization round exactly once at the declared result scale, and add ordinary/window/DISTINCT public regressions around this boundary.

2. **[P1 compatibility] Valid MatrixOne high-scale DECIMAL input is silently narrowed.**

   `AvgReturnType` caps promoted Decimal128/256 results at scale 30 (`sumavg2.go:51-58`). MatrixOne currently accepts and tests `DECIMAL(38,38)`, but a singleton exact-head probe changes

   `0.12345678901234567890123456789012345678`

   into

   `0.123456789012345678901234567890`.

   This violates the basic singleton invariant `AVG(x) = x` and regresses behavior that the base implementation preserved. MySQL-compatible inputs can follow MySQL's cap, but MO-supported extensions must not silently discard source fractional digits. Please cover ordinary/window/CTAS behavior for the supported high-scale boundary.

3. **[P1 contract] Direct integer literals still do not get MySQL-compatible metadata.**

   The updated expected result itself asserts `AVG(2)[3,23,4]` (`func_aggr_avg.result:213-226`). MySQL derives AVG precision from the argument's decimal precision, so literal `2` yields `DECIMAL(5,4)`, not the full BIGINT-domain `DECIMAL(23,4)`: https://github.com/mysql/mysql-server/blob/trunk/sql/item_sum.cc and https://github.com/mysql/mysql-server/blob/trunk/sql/item.h. Because `AvgReturnType` only sees `types.Type`, `SELECT AVG(2)`/CTAS still miss the metadata compatibility goal of #24436. Please preserve the distinction between a literal, a column, and an explicitly cast integer domain; the planner already has literal-sensitive integer typing for set operations that may be reusable.

4. **[P1 performance] The common INT32 AVG operator path regresses materially without a production-path guardrail.**

   On this exact head, a 4,096-row/64-group `BatchFill + Flush` benchmark comparing the new production `makeSumAvgExec(INT32)` path against the exact pre-PR INT32 float executor gives median `22.786 us/op` versus `12.567 us/op` (**+81.3%**); allocations remain `8 B/op, 1 alloc/op`. Exact decimal semantics may cost more, but this hot-path increase needs a production-factory benchmark and either optimization (for example, a dedicated exact integer fast path) or representative query-level evidence plus explicit acceptance of the tradeoff.

I did not find a new goroutine/resource-lifecycle or hang/deadlock issue in this change. The blockers are value correctness, schema/protocol compatibility, and the measured aggregate hot-path regression.

File: pkg/sql/colexec/aggexec/sumavg2.go [hunk 1]
```diff
@@ -32,38 +32,30 @@ import (
...
-		//4 digits.
-		return types.New(types.T_decimal128, 38, 4)
+	typ := typs[0]
+	switch typ.Oid {
+	case types.T_int8, types.T_uint8:
+		return types.New(types.T_decimal128, 7, 4)
+	case types.T_int16, types.T_uint16:
+		return types.New(types.T_decimal128, 9, 4)
+	case types.T_int32, types.T_uint32:
+		return types.New(types.T_decimal128, 14, 4)
```
Comment (aptend) [inline pkg/sql/colexec/aggexec/sumavg2.go]: [P1] Round exact AVG only once at the declared result scale. The new scale-4 integer result exposes the existing two-step finalizer: `decAvg` first calls `Decimal128.Div` (which already rounds to its selected intermediate scale) and then `Scale` rounds again. A focused exact-head probe with sum=1/count=113 produced intermediate `0.008850` and final `0.0089`, but exact `1/113` rounded directly to four places is `0.0088`. The decimal case `0.01/113` likewise returns `0.000089` rather than `0.000088` at scale 6. This affects ordinary, DISTINCT, and window AVG because all flush through `decAvg`. Please divide with sufficient unrounded precision/direct target-scale semantics and add these boundary cases.

File: pkg/sql/colexec/aggexec/sumavg2.go [hunk 2]
```diff
@@ -32,38 +32,30 @@ import (
...
+	case types.T_int64:
+		return types.New(types.T_decimal128, 23, 4)
```
Comment (aptend) [inline pkg/sql/colexec/aggexec/sumavg2.go]: [P1] Preserve expression precision instead of deriving every integer AVG from the storage OID. MySQL's `Item_sum_avg::resolve_type` uses `args[0]->decimal_precision() + div_precincrement`; integer literal `2` therefore yields `DECIMAL(5,4)`, not the blanket int64 `DECIMAL(23,4)`. The updated BVT currently records `AVG(2)[3,23,4]`, so CTAS and client/JDBC metadata remain incompatible for literals and constant expressions—the exact scenario this PR claims to align. The planner/binder needs to retain/propagate expression precision into AVG type resolution, with protocol and CTAS regression tests.

File: pkg/sql/colexec/aggexec/sumavg2.go [hunk 3]
```diff
@@ -32,38 +32,30 @@ import (
...
-			s = typs[0].Scale
-		}
-		if s > typs[0].Scale+6 {
-			s = typs[0].Scale + 6
+		precision := typ.Width + 4
+		if precision <= 38 {
+			return types.New(types.T_decimal128, precision, typ.Scale+4)
 		}
-		return types.New(types.T_decimal128, 38, s)
+		return types.New(types.T_decimal256, min(precision, 65), min(typ.Scale+4, 30))
```
Comment (aptend) [inline pkg/sql/colexec/aggexec/sumavg2.go]: [P1] Do not silently reduce a supported input's scale. `DECIMAL(38,38)` is valid in MatrixOne and the pre-PR AVG preserved scale 38, but this branch now returns `DECIMAL256(42,30)`. With a singleton input `0.12345678901234567890123456789012345678`, the exact-head finalizer returns `0.123456789012345678901234567890`, losing eight digits even though averaging one row must preserve the value and Decimal256 can represent the original scale. A MySQL compatibility cap cannot silently regress MatrixOne's wider valid domain; please define a compatibility policy that preserves supported inputs (or otherwise handles them explicitly) and cover singleton/max-scale boundaries.

File: pkg/sql/colexec/aggexec/sumavg2.go [hunk 4]
```diff
@@ -1435,20 +1468,41 @@ func makeSumAvgExec(
...
 		return newSumAvgExec[int64, int16](mp, int64OfCheck, isSum, aggID, isDistinct, param)
 	case types.T_year:
-		return newSumAvgExec[int64, int16](mp, int64OfCheck, isSum, aggID, isDistinct, param)
+		if !isSum {
+			return newSumAvgDecExec[types.MoYear, types.Decimal128](mp, isSum, aggID, isDistinct, param)
+		}
+		return newSumAvgExec[int64, types.MoYear](mp, int64OfCheck, isSum, aggID, isDistinct, param)
 	case types.T_int32:
+		if !isSum {
+			return newSumAvgDecExec[int32, types.Decimal128](mp, isSum, aggID, isDistinct, param)
```
Comment (aptend) [inline pkg/sql/colexec/aggexec/sumavg2.go]: [P1] This changes every INT32 AVG from the compact integer/float path to the generic Decimal128 executor without a performance acceptance or guardrail. On this exact head, a focused 4,096-row/64-group BatchFill+Flush benchmark (`-benchtime=200x -count=5`) measured 42.9–44.5 µs/op here versus 20.1–20.7 µs/op for the pre-PR executor—about 2.1x slower—with an extra allocation. AVG is a core aggregation hot path, so please provide an optimized exact implementation or explicit performance acceptance backed by representative benchmark coverage before landing.

**Review by aptend — CHANGES_REQUESTED** [review]
Deep review completed on exact head `adde9798bc690853fa102394907551758a477824`. I read the complete diff and PR/review/comment history; there are no inline threads or author replies to reconcile at this head. The current CI is green, but the exact-result and metadata contract still has three correctness gaps, plus a material unaccounted hot-path regression:

1. Exact AVG is rounded twice, producing wrong half-boundary results (for example, integer `1/113` becomes `0.0089` instead of `0.0088` at scale 4).
2. A valid MatrixOne `DECIMAL(38,38)` singleton is narrowed to scale 30 and loses eight fractional digits.
3. Literal precision is discarded: this PR's own BVT records `AVG(2)` as `DECIMAL(23,4)`, while MySQL derives `DECIMAL(5,4)`.
4. The integer AVG execution change is about 2.1x slower in a focused BatchFill+Flush benchmark and has no documented acceptance or guardrail.

Local validation on the exact head:
- `go test ./pkg/sql/colexec/aggexec ./pkg/sql/colexec/table_function ./pkg/sql/plan ./pkg/sql/colexec/window -count=1` — pass
- `go vet` on the same owning packages — pass
- Temporary counterexample tests — reproduced `0.0089`, `0.000089`, and high-scale singleton truncation
- `BenchmarkReview27873Int32AvgPath`, `-benchtime=200x -count=5` — current 42.9–44.5 µs/op vs pre-PR 20.1–20.7 µs/op

Please fix the blockers and add regression coverage for the counterexamples before re-requesting review.

**Review by aunjgr — APPROVED** [review]
Reviewed exact head `adde9798bc690853fa102394907551758a477824`. No material blockers found. The return-type mapping applies the MySQL exact-numeric precision/scale contract consistently, promotes wide Decimal128 AVG state to Decimal256 before division, preserves float and variance/stddev behavior, and adapts the fulltext consumer explicitly. Ordinary, window, DISTINCT, merge, NULL, overflow, CTAS, and public BVT boundaries are covered, and exact-head CI is fully green.

**Comment by daviszhen** [comment]
Addressed the requested AVG review blockers in `a0e597335e` and merged the latest `upstream/main` in `c6718adc41`:

- Finalize exact AVG by scaling the numerator to the declared result scale before one division, removing the previous double-rounding cases (`1/113 -> 0.0088`, `0.01/113 -> 0.000088`), with ordinary/DISTINCT/window regressions.
- Preserve supported high-scale DECIMAL input, including `DECIMAL(38,38) -> DECIMAL(42,38)`, with ordinary/window execution and CTAS metadata coverage.
- Keep direct integer literal metadata value-sensitive (`AVG(2) -> DECIMAL(5,4)`), while columns and explicit integer casts retain their full domains.
- Add a production-factory INT32 AVG benchmark plus a dedicated native integer finalization path; the local 4,096-row/64-group benchmark is now close to the pre-PR float reference.
- Reduced the binder change to the direct-literal case and added coverage for the new branches; updated affected distributed metadata fixtures.

Validation on macOS with the repository third-party headers:
`go test ./pkg/sql/colexec/aggexec ./pkg/sql/plan ./pkg/sql/colexec/window -count=1`
`go vet ./pkg/sql/colexec/aggexec ./pkg/sql/plan ./pkg/sql/colexec/window`
Both pass. Please re-review the new head.

File: pkg/sql/colexec/aggexec/sumavg2.go (hunk 1 above)
Comment (daviszhen) [reply pkg/sql/colexec/aggexec/sumavg2.go]: Fixed in d01073c04e. AVG now scales the numerator to the declared result scale before one division, covering positive/negative integer and decimal boundary cases in ordinary, DISTINCT, and window paths.

File: pkg/sql/colexec/aggexec/sumavg2.go (hunk 2 above)
Comment (daviszhen) [reply pkg/sql/colexec/aggexec/sumavg2.go]: Fixed in d01073c04e. The binder now records precision only for direct integer literals (including unary +/- and parentheses), while integer columns and explicit integer casts keep their full storage-domain precision. AVG(2) is DECIMAL(5,4); cast/column metadata remains wide. CTAS and window metadata regressions are covered.

File: pkg/sql/colexec/aggexec/sumavg2.go (hunk 3 above)
Comment (daviszhen) [reply pkg/sql/colexec/aggexec/sumavg2.go]: Fixed in d01073c04e. AVG decimal scale is capped at 38 but never below the input scale, so DECIMAL(38,38) returns DECIMAL256(42,38) and preserves a singleton value exactly. Ordinary, DISTINCT, window, and CTAS cases are covered.

File: pkg/sql/colexec/aggexec/sumavg2.go (hunk 4 above)
Comment (daviszhen) [reply pkg/sql/colexec/aggexec/sumavg2.go]: Addressed in d01073c04e. Native integer AVG remains on the compact accumulator path and uses a dedicated bits.Mul64/bits.Div64 fast path for the common INT32 case; non-DISTINCT results are appended in one fixed list. Added paired production-vs-pre-PR benchmark coverage; local runs are in the same low-teens microsecond range with allocations reduced from 66 to 4.

**Comment by daviszhen** [comment]
Follow-up pushed in 23cff803a6 (on top of d01073c04e). The coverage gate was reproduced locally with the CI parser: 65.61% on the first revision; after removing unreachable duplicate branches and adding exact-AVG boundary/conversion coverage, it is 76.61% (>75%). Local aggexec/plan/window tests and vet pass; the paired INT32 benchmark remains within low-teens microseconds (exact 12.1–14.2 us/op vs float reference 11.8–12.4 us/op, 4 vs 1 allocs). New CI run: https://github.com/matrixorigin/matrixone/actions/runs/33599522728.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
Deep re-review of exact head `af506cc077f9d29ce9ceee727cce8d5ed78090cd` against base `c46d897e9645b80178568ef0783dd8e99e527222`, including the complete AVG type/execution change, the increment after my prior requested-changes head `adde9798bc69`, ordinary/DISTINCT/window/distributed merge paths, negative rounding, Decimal128-to-256 promotion, high-scale inputs, allocation/error cleanup, internal consumers, and constant-expression counterexamples.

The four prior blockers are materially addressed: exact AVG now rounds once at the declared scale; valid high-scale Decimal128 input is promoted without truncating its source scale; direct literal metadata is value-sensitive; and the common INT32 path has a dedicated production fast finalizer plus benchmark coverage. I found no new lifecycle, deadlock, or common-path performance blocker. One correctness/type-system boundary remains:

1. **[P1][correctness/metadata] Integer constant-expression precision is unbounded, but integer AVG is always declared as Decimal128.** `avgIntegerConstantPrecision` recursively adds precision for multiplication/addition (`base_binder.go:3636-3665`) and writes it directly into the integer argument type. `AvgReturnType` then unconditionally returns `DECIMAL128(precision+4,4)` for every integer OID (`sumavg2.go:35-40`) without the Decimal128 limit/promotion used by decimal inputs.

   A valid, non-overflowing counterexample is:
   ```sql
   SELECT AVG(0 * 100000000 * 100000000 * 100000000 * 100000000);
   ```
   The expression value is exactly zero, while the new AST rule derives precision `1+9+9+9+9 = 37`, so AVG is declared as `DECIMAL128(41,4)`. That width is not representable by the Decimal128 public type; CTAS and protocol metadata can therefore expose an invalid schema even though execution data is trivial. Addition/cancellation chains can reach the same boundary.

   This must be closed coherently, not by changing only the return OID: `flushExactAvg` currently states that native integer AVG always returns Decimal128, and its non-Decimal128 branch calls `decAvg[Decimal128]`, which rejects a Decimal256 result type (`sumavg2.go:873-917`). Define a bounded type rule for integer constant expressions and make planner, factory/state, ordinary/DISTINCT/window finalization, CTAS, and protocol metadata agree. Add boundary tests immediately below/above precision 38, including a zero/cancellation expression so value overflow cannot mask the metadata defect.

Validation note: exact-head substantive CI is terminal green and current main has no post-base overlap in the touched production paths. Focused local compilation was blocked by this isolated worktree's missing third-party `usearch.h` while building its exact CGo prerequisite, so I do not claim a local test pass; the blocker above follows directly from the exact production type equations and existing test-confirmed integer-expression path.

**Comment by daviszhen** [comment]
Addressed the remaining wide integer constant-expression AVG boundary in commit `0d4d0c90ef`:

- Bound integer constant-expression precision to the public `Decimal256(65,4)` AVG contract.
- Promote AVG from `Decimal128` to `Decimal256` at input precision 35 (result precision 39), while keeping precision 34 at `Decimal128(38,4)`.
- Added Decimal256 native-integer finalization for ordinary and DISTINCT paths; the Decimal256 state factory is selected for wide int64/uint64 AVG, and window finalization remains consistent.
- Added zero-expression CTAS metadata coverage for precision 34/35/37 and execution coverage for ordinary/DISTINCT/window modes.

Validation:
- `go test ./pkg/sql/colexec/aggexec ./pkg/sql/plan -count=1`
- `go test ./pkg/sql/colexec/window -count=1`
- `go vet ./pkg/sql/colexec/aggexec ./pkg/sql/plan`
- `make build` (passed; only macOS duplicate-library linker warnings)

**Comment by daviszhen** [comment]
Follow-up: the branch was amended to `f29e4b2064` to clarify the AVG precision comment; the implementation and validation described above are unchanged.

**Comment by daviszhen** [comment]
Follow-up for the coverage check: commit `14557369db` adds coverage for the wide integer AVG Decimal256 native finalizer (int32 ordinary/DISTINCT/window) and the planner precision cap above 61 digits. The previous 74.01% modified-line coverage failure is addressed; local aggregate/plan tests pass.

**Comment by daviszhen** [comment]
CI run [33646744442](https://github.com/matrixorigin/matrixone/actions/runs/33646744442) is now green. The Ubuntu UT, SCA, both BVT jobs, UT Coverage, and final Coverage merge all passed on head `14557369db`. The previous 74.01% coverage failure is resolved.

**Comment by XuPeng-SH** [comment]
当前 PR 与最新 main 存在 merge conflict，暂无法进入 review 队列。请 rebase 最新 main 并解决冲突后重新 push，完成后我再继续 review。

**Review by XuPeng-SH — APPROVED** [review]
